libqtile/widget/prompt.py

```diff
@@ -403,36 +431,109 @@ def _update(self):
             self.text = ""
         self.bar.draw()
 
+    def _trigger_complete(self):
+        # Trigger the autocompletion in user input
+        self.userInput = self.completer.complete(self.userInput)
+
+    def _history_to_input(self):
+        # Move actual command (when exploring history) to user input and update
+        # history position (right after the end)
+        if self.archivedInput:
+            self.userInput = self.archivedInput
+            self.archivedInput = ""
+            self.position = len(self.history)
+
+    def _write_char(self):
+        # Add pressed (legal) char key to user input.
+        # No LookupString in XCB... oh, the shame! Unicode users beware!
+        self._history_to_input()
+        self.userInput += chr(self.key)
+        del self.key
+
+    def _backspace(self):
+        # Delete the last char from user input
+        self._history_to_input()
+        if len(self.userInput) > 0:
+            self.userInput = self.userInput[:-1]
+
+    def _unfocus(self):
+        # Remove focus from the widget
+        self.active = False
+        self.bar.widget_ungrab_keyboard()
+
+    def _send_cmd(self):
+        # Send the prompted text for execution
+        self._unfocus()
+        if self.strict_completer:
+            self.userInput = self.actual_value or self.userInput
+            del self.actual_value
+        self._history_to_input()
+        if self.userInput:
+            # If history record is activated, also save command in history
+            if self.record_history:
+                self.history.append(self.userInput)
+                if self.position < self.max_history:
+                    self.position += 1
+                with open(self.history_path, mode='wb') as f:
+                    pickle.dump(self.history, f, protocol=2)
+            self.callback(self.userInput)
+
+    def _get_prev_cmd(self):
+        # Get the previous command in history.
+        # If there isn't more previous commands, ring system bell
+        if self.record_history:
+            if not self.position:
+                print("\a")
+            else:
+                self.position -= 1
+                self.archivedInput = self.history[self.position]
+
+    def _get_next_cmd(self):
+        # Get the next command in history.
+        # If the last command was already reached, ring system bell.
+        if self.record_history:
+            if self.position == len(self.history):
+                print("\a")
```

Comment:
I think it probably depends on how your login manager spawns things; you may not be connected to a tty at all. You can use `os.isatty` to detect this if you really want to keep the beep, though.